// flagsmith-nodejs/build/cjs/sdk/polling_manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnvironmentDataPollingManager = void 0;
class EnvironmentDataPollingManager {
    interval;
    main;
    refreshIntervalSeconds;
    logger;
    constructor(main, refreshIntervalSeconds, logger) {
        this.main = main;
        this.refreshIntervalSeconds = refreshIntervalSeconds;
        this.logger = logger;
    }
    start() {
        const updateEnvironment = () => {
            if (this.interval)
                clearInterval(this.interval);
            this.interval = setInterval(async () => {
                try {
                    await this.main.updateEnvironment();
                }
                catch (error) {
                    this.logger.error(error, 'failed to poll environment');
                }
            }, this.refreshIntervalSeconds * 1000);
        };
        updateEnvironment();
    }
    stop() {
        if (!this.interval) {
            return;
        }
        clearInterval(this.interval);
    }
}
exports.EnvironmentDataPollingManager = EnvironmentDataPollingManager;
